contracts/pools/aeth/StableSwapAETH.vy

```diff
@@ -0,0 +1,857 @@
+    """
+    amp: uint256 = self._A()
+    rates: uint256[N_COINS] = self._stored_rates()
+    _balances: uint256[N_COINS] = self.balances
+    D0: uint256 = self.get_D_mem(rates, _balances, amp)
+    for i in range(N_COINS):
+        _amount: uint256 = amounts[i]
+        if is_deposit:
+            _balances[i] += _amount
+        else:
+            _balances[i] -= _amount
+    D1: uint256 = self.get_D_mem(rates, _balances, amp)
+    token_amount: uint256 = ERC20(self.lp_token).totalSupply()
+    diff: uint256 = 0
+    if is_deposit:
+        diff = D1 - D0
+    else:
+        diff = D0 - D1
+    return diff * token_amount / D0
+
+@payable
+@external
+@nonreentrant('lock')
+def add_liquidity(amounts: uint256[N_COINS], min_mint_amount: uint256) -> uint256:
+    """
+    @notice Deposit coins into the pool
+    @param amounts List of amounts of coins to deposit
+    @param min_mint_amount Minimum amount of LP tokens to mint from the deposit
+    @return Amount of LP tokens received by depositing
+    """
+    assert not self.is_killed
+
+    rates: uint256[N_COINS] = self._stored_rates()
+    _lp_token: address = self.lp_token
+    token_supply: uint256 = ERC20(_lp_token).totalSupply()
+
+    # Initial invariant
+    amp: uint256 = self._A()
+    D0: uint256 = 0
+    old_balances: uint256[N_COINS] = self.balances
+    if token_supply != 0:
+        D0 = self.get_D_mem(rates, old_balances, amp)
+
+    new_balances: uint256[N_COINS] = old_balances
+    for i in range(N_COINS):
+        if token_supply == 0:
+            assert amounts[i] > 0
+        new_balances[i] = old_balances[i] + amounts[i]
+
+    # Invariant after change
+    D1: uint256 = self.get_D_mem(rates, new_balances, amp)
+    assert D1 > D0
+
+    # We need to recalculate the invariant accounting for fees
+    # to calculate fair user's share
+    D2: uint256 = D1
+    fees: uint256[N_COINS] = empty(uint256[N_COINS])
+    if token_supply != 0:
+        # Only account for fees if we are not the first to deposit
+        _fee: uint256 = self.fee * N_COINS / (4 * (N_COINS - 1))
+        _admin_fee: uint256 = self.admin_fee
+        for i in range(N_COINS):
+            ideal_balance: uint256 = D1 * old_balances[i] / D0
+            difference: uint256 = 0
+            if ideal_balance > new_balances[i]:
+                difference = ideal_balance - new_balances[i]
+            else:
+                difference = new_balances[i] - ideal_balance
+            fees[i] = _fee * difference / FEE_DENOMINATOR
+            self.balances[i] = new_balances[i] - (fees[i] * _admin_fee / FEE_DENOMINATOR)
+            new_balances[i] -= fees[i]
+        D2 = self.get_D_mem(rates, new_balances, amp)
+    else:
+        self.balances = new_balances
+
+    # Calculate, how much pool tokens to mint
+    mint_amount: uint256 = 0
+    if token_supply == 0:
+        mint_amount = D1  # Take the dust if there was any
+    else:
+        mint_amount = token_supply * (D2 - D0) / D0
```

Comment:
This section can be merged with the previous one to avoid 2 checks on `token_supply == 0`.  See one of the latest pools for an example.